modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/validators.js

```diff
@@ -0,0 +1,366 @@
+ */
+function composeOneOfErrorMessage(arrayOfValues) {
+	return `${ERROR_ONE_OF} ${JSON.stringify(arrayOfValues)}.`;
+}
+
+/**
+ * Returns the type of the given value.
+ * @param {*} value Any value.
+ * @return {string} Type of value.
+ */
+function getType(value) {
+	return Array.isArray(value) ? 'array' : typeof value;
+}
+
+/**
+ * Checks if the given validator result says that the value is invalid.
+ * @param {boolean|!Error} result
+ * @return {boolean}
+ */
+function isInvalid(result) {
+	return result instanceof Error;
+}
+
+/**
+ * Wraps the given validator so that it also accepts null/undefined values.
+ *   a validator that checks a value against a single type, null, or
+ * undefined.
+ * @param {!function()} typeValidator Validator to wrap.
+ * @return {!function()} Wrapped validator.
+ */
+function maybe(typeValidator) {
+	return (value, name, context) => {
+		return _isDefAndNotNull(value)
+			? typeValidator(value, name, context)
+			: true; // eslint-disable-line
+	};
+}
+
+/**
+ * Throws error if validator is invoked with incorrect type.
+ * @param {string} expectedType String representing the expected type.
+ * @param {*} value The value to match the type of.
+ * @param {!string} name Name of the function the validator is intended for.
+ */
+function throwConfigError(expectedType, value, name) {
+	const type = getType(value);
+	throw new Error(
+		`Expected type ${expectedType}, but received type ${type}. passed to ${name}.`
+	);
+}
+
+/**
+ * Checks if all the items of the given array pass the given validator.
+ * @param {!function()} validator
+ * @param {*} value The array to validate items for.
+ * @param {string} name The name of the array property being checked.
+ * @param {!Object} context Owner of the array property being checked.
+ * @return {!Error|boolean} `true` if the type matches, or an error otherwise.
+ */
+function validateArrayItems(validator, value, name, context) {
+	for (let i = 0; i < value.length; i++) {
+		if (isInvalid(validator(value[i], name, context))) {
+			const itemValidatorError = validator(value[i], name, context);
+			const errorMessage = `Validator for ${name}[${i}] says: "${itemValidatorError}"`;
+
+			return composeError(errorMessage, name, context);
+		}
+	}
+
+	return true;
+}
+
+/**
+ * Checks if the given value matches the expected type.
+ * @param {string} expectedType String representing the expected type.
+ * @param {*} value The value to match the type of.
+ * @param {string} name The name of the property being checked.
+ * @param {!Object} context Owner of the property being checked.
+ * @return {!Error|boolean} `true` if the type matches, or an error otherwise.
+ */
+function validateType(expectedType, value, name, context) {
+	const type = getType(value);
+	if (type !== expectedType) {
+		const msg = `Expected type '${expectedType}', but received type '${type}'.`;
+
+		return composeError(msg, name, context);
+	}
+
+	return true;
+}
+
+export default validators;
```

Comment:
Do we need all of these validators? If there are any that aren't actually used, now would be a good time to prune them, so that we can reduce our API surface area and download bytes.

Speaking of download bytes, better to do individual exports instead of export all of them as a single object. This is why JS allows multiple `export` statements in the first place. For potential tree-shaking purposes (even though in reality we won't _practically_ tree-shake these, it would be nice to be able, _theoretically_, to tree-shake them in the future).